File: src/commands/commands.ts
import exec from 'exec';

const cmdLinux = 'sudo systemctl';
const cmdMac = 'brew services';

type Status = {
  service: string,
  status: 'online' | 'offline' | 'loading'
}

type Output = {
  stdout: string,
  stderr: string
}

// Using an interface to avoid using computed enums in TS
interface Commands {
  serviceStart(service: string): Promise<Output>,
  serviceStop(service: string): Promise<Output>,
  serviceRestart(service: string): Promise<Output>,
  serviceStatus(service: string): Promise<Status>
}

let commands: Commands;

const macServiceStatus = async (service: string) => {
  let output = await exec(`brew services list | grep ${service}`);

  let splitLine = output.stdout.replace(/\s+/g, ' ').split(' ');
  const statusString = splitLine[1];

  if (statusString.trim() === 'started') {
    const status: Status = {
      service,
      status: 'online'
    }
    return status;
  }

  const status: Status = {
    service,
    status: 'offline'
  }
  return status;
}

const linuxServiceStatus = async (service: string) => {
  let output = await exec(`sudo systemctl status ${service} | grep Active`);

  if (output.stdout.includes('active (running')) {
    const status: Status = {
      service,
      status: 'online'
    }
    return status;
  }

  if (output.stdout.includes('active (waiting')) {
    const status: Status = {
      service,
      status: 'loading'
    }
    return status;
  }

  const status: Status = {
    service,
    status: 'offline'
  }
  return status;
}

const macStartService = async (service: string) => {
  let output = await exec(`brew services start ${service}`);

  return output;
}

const linuxStartService = async (service: string) => {
  let output = await exec(`sudo systemctl start ${service}`);

  return output;
}

const macStopService = async (service: string) => {
  let output = await exec(`brew services stop ${service}`);

  return output;
}

const linuxStopService = async (service: string) => {
  let output = await exec(`sudo systemctl stop ${service}`);

  return output;
}

const macRestartService = async (service: string) => {
  let output = await exec(`brew services restart ${service}`);

  return output;
}

const linuxRestartService = async (service: string) => {
  let output = await exec(`sudo systemctl restart ${service}`);

  return output;
}

if (process.platform === 'darwin') {
  commands = {
    serviceStart: macStartService,
    serviceStop: macStopService,
    serviceRestart: macRestartService,
    serviceStatus: macServiceStatus
  }
} else {
  commands = {
    serviceStart: linuxStartService,
    serviceStop: linuxStopService,
    serviceRestart: linuxRestartService,
    serviceStatus: linuxServiceStatus
  }
}

export default commands;

File: src/commands/commands.rs
use std::process::Command;
pub struct Commands;

enum Status {
  Online,
  Offline,
  Loading
}

pub struct SysService {
  name: String,
  status: Status
}

impl Commands {
  fn command_prefix() -> String {
    if std::env::consts::OS == "macos" {
      return "brew services".to_owned();
    }

    return "sudo systemctl".to_owned()
  }

  pub fn service_start(service: &str) -> Command {
    let command = format!("{} start {}", Commands::command_prefix(), &service);

    Command::new(command)
  }

  pub fn service_stop(service: &str) -> Command {
    let command = format!("{} stop {}", Commands::command_prefix(), &service);

    Command::new(command)
  }

  pub fn service_restart(service: &str) -> Command {
    let command = format!("{} restart {}", Commands::command_prefix(), &service);

    Command::new(command)
  }

  pub fn service_status(service: &str) -> SysService {
    let command = format!("{} status {}", Commands::command_prefix(), &service);

    let mut command_get_status = "grep Active".to_owned();

    if std::env::consts::OS == "macos" {
      command_get_status = format!("grep {}", &service);
      let command_string = format!("{} | {}", command, command_get_status);

      let output = Command::new(command_string).output().expect("Failed to execute.");
      let stdout = String::from_utf8(output.stdout).expect("Invalid UTF-8");

      if stdout.contains("started") {
        return SysService {
          name: service.to_owned(),
          status: Status::Online
        }
      }

      return SysService {
        name: service.to_owned(),
        status: Status::Offline
      }
    }

    let command_string = format!("{} | {}", command, command_get_status);
    let output = Command::new(command_string).output().expect("Failed to execute");
    let stdout = String::from_utf8(output.stdout).expect("Invalid UTF-8");

    if stdout.contains("active (running)") {
      return SysService {
        name: service.to_owned(),
        status: Status::Online
      }
    }

    if stdout.contains("active (waiting)") {
      return SysService {
        name: service.to_owned(),
        status: Status::Loading
      }
    }

    return SysService {
      name: service.to_owned(),
      status: Status::Offline
    }
  }
}
